server/server/src/billing/payment_service.rs

```diff
@@ -0,0 +1,449 @@
+        &subscription_resp.id,
+        subscription_resp.current_period_end,
+    )
+    .await
+    .map_err(|e| {
+        InternalError(format!(
+            "Cannot add stripe subscription in Postgres: {:?}",
+            e
+        ))
+    })?;
+
+    file_index_repo::set_account_data_cap(&mut transaction, &public_key, PAID_TIER_SIZE)
+        .await
+        .map_err(|e| {
+            InternalError(format!(
+                "Cannot change user data cap to premium data cap: {:?}",
+                e
+            ))
+        })
+}
+
+pub async fn get_last_registered_credit_card(
+    context: RequestContext<'_, GetLastRegisteredCreditCardRequest>,
+) -> Result<GetLastRegisteredCreditCardResponse, ServerError<GetLastRegisteredCreditCardError>> {
+    let mut transaction = match context.server_state.index_db_client.begin().await {
+        Ok(t) => t,
+        Err(e) => {
+            return Err(InternalError(format!("Cannot begin transaction: {:?}", e)));
+        }
+    };
+
+    let credit_card =
+        file_index_repo::get_last_stripe_credit_card_info(&mut transaction, &context.public_key)
+            .await
+            .map_err(|e| match e {
+                GetLastStripeCreditCardInfoError::NoPaymentInfo => {
+                    ClientError(GetLastRegisteredCreditCardError::OldCardDoesNotExist)
+                }
+                _ => InternalError(format!("Cannot get all stripe credit card infos: {:?}", e)),
+            })?;
+
+    Ok(GetLastRegisteredCreditCardResponse {
+        credit_card_last_4_digits: credit_card.last_4_digits,
+    })
+}
+
+pub enum StripeWebhookError {
+    VerificationError(String),
+    InvalidHeader(String),
+    InvalidBody(String),
+    InternalError(String),
+}
+
+pub async fn stripe_webhooks(
+    state: &Arc<ServerState>,
+    request_body: Bytes,
+    stripe_sig: HeaderValue,
+) -> Result<(), StripeWebhookError> {
+    let webhook = match serde_json::from_slice::<StripeWebhook>(request_body.as_ref()) {
+        Ok(webhook) => webhook,
+        Err(e) => {
+            return Err(InvalidHeader(format!(
+                "Cannot deserialize stripe webhook body: {:?}",
+                e
+            )))
+        }
+    };
+
+    let event_type = match webhook.event_type {
+        StripeMaybeContainer::Expected(ref known) => known,
+        StripeMaybeContainer::Unexpected(_) => return Ok(()),
+    };
+
+    let json = match std::str::from_utf8(&request_body.as_ref()) {
+        Ok(json) => json,
+        Err(e) => {
+            return Err(InvalidBody(format!(
+                "Cannot turn stripe webhook body bytes into str: {:?}",
+                e
+            )))
+        }
+    };
+
+    let mut sig_header_split = stripe_sig
+        .to_str()
+        .map_err(|e| {
+            InvalidHeader(format!(
+                "Cannot turn stripe webhook header into str: {:?}",
+                e
+            ))
+        })?
+        .split(",");
+
+    let time = sig_header_split
+        .next()
```

Comment:
I don't know if you're gaurenteed that the order is going to be v1 then v0, you should check to make sure you have the right values, and not depend on `next`